CO-3587 and CO-3586 taken from coup1234

ecino opened the pull request:
Replaces #1298

coup1234 committed:
CO-3586: Add new partner automatically to mailchimp:
- At partner creation, a new mass_mailing_contact is created and
exported to all existing mailchimp List if email is provided and
no optout.
- Same when the email is provided later

coup1234 committed:
CO-3587: Many2Many relation between partner and mailing contact:
- Merged mailchimp field are now also in mass_mailing contact (mailchimp
 module needs to be updated)
- Ensure notification of mailchimp and mass_mailing contact in case of
any changes in the partner.
- Update script for copy old partner_id in partner_ids and compute other
fields

coup1234 committed:
CO-3587: Syncronize tags beteweenn partner, mass_mailing_contact and mailchimp. - Improve partner deletion and modification

coup1234 committed:
CO-3587: Rebase from devel and bug fix

JonathanGuerne commented on mass_mailing_switzerland/__manifest__.py:
```diff
@@ -47,6 +47,7 @@
         "views/account_invoice_line_view.xml",
         "views/mass_mailing_contact_view.xml",
         "views/mass_mailing_settings_view.xml",
+        "views/update.xml",
```
use a migration instead

JonathanGuerne commented on mass_mailing_switzerland/models/mass_mailing_contact.py:
```diff
@@ -14,27 +14,224 @@
+            if i % 10 == 0:
+                print(i)
+            if not rec.partner_ids:
+                rec.add_partners(rec.partner_id.id)
```
If we want already existing partner with duplicated email to be merge on a single mailing contact the logic should be added here. If we do so some contact could end up useless at the end of the process. They could be archived

JonathanGuerne commented on mass_mailing_switzerland/models/mass_mailing_contact.py:
```diff
@@ -14,27 +14,224 @@
+                salutation = title_salutation + " " + family_title.name + " " + lang_partner.lastname
+            else:
+                partners = partner_ids.with_context(lang=partner_ids[0].lang)
+                if partners.search([("title.gender", "=", "M")]):
```
Use filtered instead of search

JonathanGuerne commented on mass_mailing_switzerland/models/mass_mailing_contact.py:
```diff
@@ -14,27 +14,224 @@
+            if contact:
+                contact.add_partners(vals_list['partner_id'])
+                contact.action_update_to_mailchimp()
+                return None
```
could cause some issue if caller expect a mailling.contact

JonathanGuerne commented on mass_mailing_switzerland/models/mass_mailing_contact.py:
```diff
@@ -14,27 +14,224 @@
+    def _prepare_vals_for_merge_fields(self, mailchimp_list_id):
+        merge_fields_vals = {}
+        partner_id = self.partner_id
+        test = mailchimp_list_id.merge_field_ids
```
```suggestion

```

JonathanGuerne commented on mass_mailing_switzerland/models/mass_mailing_contact.py:
```diff
@@ -14,27 +14,224 @@
+                contact.compute_salutation()
+                contact.compute_sponsored_child_fields()
+                contact.action_update_to_mailchimp()
+        return True
```
```suggestion

```

This is duplicated

JonathanGuerne commented on mass_mailing_switzerland/models/mass_mailing_contact.py:
```diff
@@ -97,16 +299,16 @@ def update_partner_merge_fields_job(self, partner_id):
     @job(default_channel="root.mass_mailing_switzerland.update_mailchimp")
     @api.model
-    def update_all_merge_fields_job(self, partner_ids=None):
+    def update_all_merge_fields_job(self, partner_to_update=None):
```
`update_partner_merge_fields_job` has to be change. it rely on `mailing.contact` `partner_id` field

JonathanGuerne commented on mass_mailing_switzerland/models/mass_mailing_contact.py:
```diff
@@ -45,6 +242,11 @@ def get_partner(self, email):
+            for contact in self.filtered(lambda c: c.partner_ids):
+                contact.compute_salutation()
+                contact.compute_sponsored_child_fields()
+        self.action_update_to_mailchimp()
```
```suggestion
            contact.with_delay().update_all_merge_fields_job(contact.partner_ids[0].id)
```
update should be handle trough cron

JonathanGuerne commented on mass_mailing_switzerland/models/mass_mailing_contact.py:
```diff
@@ -97,16 +299,16 @@ def update_partner_merge_fields_job(self, partner_id):
+                lambda a: _filter_criteria(a.partner_ids, partner_to_update)):
+            for partner in mailing_contact.partner_ids:
+                # call update on each partner linked to each mailing contact
+                self.with_delay().update_partner_merge_fields_job(partner.id)
```
function need to be rethink. The contact is updated to mailchimp not the partner directly

JonathanGuerne commented on mass_mailing_switzerland/models/res_partner.py:
```diff
@@ -12,14 +12,20 @@
+        default_val = self.env['mailchimp.lists'].search([])
+        return default_val
+
+    odoo_list_ids = fields.Many2many('mailchimp.lists', string='MailChimp Lists', default=_default_odoo_list_ids)
```
Isn't there an other way to retrieve a partner link to a mailchimp list ?
This could be a potential duplicate

JonathanGuerne commented on mass_mailing_switzerland/models/res_partner.py:
```diff
@@ -12,14 +12,20 @@
     _inherit = "res.partner"
 
+    def _default_odoo_list_ids(self):
+        default_val = self.env['mailchimp.lists'].search([])
```
```suggestion
        default_val = self.env['mailchimp.lists'].search([], limit=1)
```

JonathanGuerne commented on mass_mailing_switzerland/models/res_partner.py:
```diff
@@ -129,61 +135,89 @@ def _compute_sponsored_child_fields(self):
+                if not partner.opt_out:
+                    vals["partner_id"] = partner.id
+                    vals["odoo_list_ids"] = partner.odoo_list_ids
+                    mass_mailing_contact = self.env["mail.mass_mailing.contact"].create(vals)
+                    mass_mailing_contact.name = self.firstname + " " + self.lastname
+                    mass_mailing_contact.action_export_to_mailchimp()
```
self is a collection. this could lead to errors

name should be added in vals

export is already done on contact creation

JonathanGuerne commented on mass_mailing_switzerland/views/update.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<odoo noupdate="1">
+   <data>
+    <function model="mail.mass_mailing.contact" name="move_m2o_to_m2m"/>
+  </data>
+</odoo>
```
Use a migration instead

JonathanGuerne commented on sponsorship_switzerland/models/contracts.py:
```diff
@@ -387,6 +394,15 @@ def _on_sponsorship_finished(self):
+        for contact in partners.filtered(lambda c: c.mass_mailing_contact_ids):
+            for mass_mailing_contact in contact.mass_mailing_contact_ids:
+                mass_mailing_contact.compute_sponsored_child_fields()
+                mass_mailing_contact.action_update_to_mailchimp()
```
```suggestion
                mass_mailing_contact.with_delay().update_all_merge_fields_job(mass_mailing_contact.partner_ids[0].id)
```

ecino committed:
CO-3586 CO-3587 Code review

- Remove code for exporting new partners: a CRON task exist for that!
  (Auto Export Contact to MailChimp List)

JonathanGuerne commented on mass_mailing_switzerland/models/mass_mailing_contact.py:
```diff
@@ -43,80 +202,22 @@ def get_partner(self, email):
+            for contact in self:
+                if contact.id in args:
+                    to_update -= contact
+        return to_update.wit_delay().action_update_to_mailchimp()
```
```suggestion
        return to_update.with_delay().action_update_to_mailchimp()
```
